File: cpp/src/solutions/valid_parentheses.cpp
// LeetCode 20. Valid Parentheses in C++

#include "cpp/src/common/common.h"

class Solution {
 public:
  bool isValid(string s) {
    stack<char> stack;
    for (auto ch : s) {
      if (ch == '(' || ch == '{' || ch == '[') {
        stack.push(ch);
      } else if (!stack.empty() && ((ch == ')' && stack.top() == '(') ||
                                    (ch == '}' && stack.top() == '{') ||
                                    (ch == ']' && stack.top() == '['))) {
        stack.pop();
      } else {
        return false;
      }
    }
    return stack.empty();
  }
};

TEST(isValid, example1) {
  string s = "()";
  bool ans = true;
  EXPECT_EQ(Solution().isValid(s), ans);
}

TEST(isValid, example2) {
  string s = "()[]{}";
  bool ans = true;
  EXPECT_EQ(Solution().isValid(s), ans);
}

TEST(isValid, example3) {
  string s = "(]";
  bool ans = false;
  EXPECT_EQ(Solution().isValid(s), ans);
}

TEST(isValid, example4) {
  string s = "([)]";
  bool ans = false;
  EXPECT_EQ(Solution().isValid(s), ans);
}

TEST(isValid, example5) {
  string s = "{[]}";
  bool ans = true;
  EXPECT_EQ(Solution().isValid(s), ans);
}

File: rust/src/solutions/valid_parentheses.rs
// LeetCode 20. Valid Parentheses

use super::Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
        for ch in s.chars() {
            match ch {
                '(' | '{' | '[' => {
                    stack.push(ch);
                }
                ')' if stack.last() == Some(&'(') => {
                    stack.pop();
                }
                '}' if stack.last() == Some(&'{') => {
                    stack.pop();
                }
                ']' if stack.last() == Some(&'[') => {
                    stack.pop();
                }
                _ => return false,
            }
        }
        stack.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s = String::from("()");
        let ans = true;
        assert_eq!(Solution::is_valid(s), ans);
    }

    #[test]
    fn example2() {
        let s = String::from("()[]{}");
        let ans = true;
        assert_eq!(Solution::is_valid(s), ans);
    }

    #[test]
    fn example3() {
        let s = String::from("(]");
        let ans = false;
        assert_eq!(Solution::is_valid(s), ans);
    }

    #[test]
    fn example4() {
        let s = String::from("([)]");
        let ans = false;
        assert_eq!(Solution::is_valid(s), ans);
    }

    #[test]
    fn example5() {
        let s = String::from("{[]}");
        let ans = true;
        assert_eq!(Solution::is_valid(s), ans);
    }
}
